models/databasehelper.go

```diff
@@ -0,0 +1,76 @@
+package models
+
+import (
+	// "database/sql"
+	"errors"
+	"github.com/lib/pq"
+)
+
+// UniqueConstraintError is returned when a uniqueness constraint is violated during an insert.
+var UniqueConstraintError = errors.New("postgres: unique constraint violation")
+
+// QueryResultContainedMultipleRowsError is returned when a query unexpectedly returns more than one row.
+var QueryResultContainedMultipleRowsError = errors.New("query result unexpectedly contained multiple rows")
+
+// QueryResultContainedNoRowsError is returned when a query unexpectedly returns no rows.
+var QueryResultContainedNoRowsError = errors.New("query result unexpectedly contained no rows")
+
+func convertPostgresError(err error) error {
+	const uniqueConstraintErrorCode = "23505"
+
+	if postgresErr, ok := err.(*pq.Error); ok {
+		if postgresErr.Code == uniqueConstraintErrorCode {
+			return UniqueConstraintError
+		}
+	}
+
+	return err
+}
+
+func (db *DB) execOneResult(sqlQuery string, object interface{}, args ...interface{}) error {
+
+	rows, err := db.Query(sqlQuery, args...)
+	if err != nil {
+		return convertPostgresError(err)
+	}
+	defer rows.Close()
+
+	foundResult := false
+	for rows.Next() {
```

Comment:
Something bugged me a lot about this function. Took me a while to figure out what. By reading the line `for rows.Next() {` it looks like this is basically a while loop that runs until there are no more rows. You have to read the body of the loop and run the code in your head to figure out that this "loop" actually executes at most twice, so it's not really a loop.

What you really want is to just grab the first row and then assert that there are no more rows left.
```go
func (db *DB) execOneResult(sqlQuery string, object interface{}, args ...interface{}) error {
 	rows, err := db.Query(sqlQuery, args...)
	if err != nil {
		return convertPostgresError(err)
	}
	defer rows.Close()

	hasFirstRow := rows.Next()
	if !hasFirstRow {
	    return QueryResultContainedNoRowsError
	}
	
	if err := rows.Scan(object); err != nil {
	    return convertPostgresError(err)
	}
	
 	if err := rows.Err(); err != nil {
		return convertPostgresError(err)
	}
	hasAnotherRow := rows.Next()
	if hasAnotherRow {
	  return QueryResultContainedMultipleRowsError
	}

	return nil
}
```